### ISSUE TITLE
`no-invalid-this` should allow `this` in assigned methods

### ISSUE THREAD

**Issue by errorx666** [palantir/tslint#1544]
### Bug Report
- **TSLint version**: `3.15.1`
- **TypeScript version**: `2.1.0-dev.2016-0907`
- **Running TSLint via**: gulp
#### TypeScript code being linted

``` ts
const systemLocate = System.locate,
    suffix = `?nocache=${Math.random()}`;
System.locate = function( ...args ) {
    return Promise.resolve( systemLocate.apply( this, args ) )
        .then( addr => addr + suffix );
};
```

with `tslint.json` configuration:

``` json
{
"no-invalid-this": [
            true,
            "check-function-in-method"
        ]
}
```
#### Actual behavior

`the "this" keyword is disallowed outside of a class body`
#### Expected behavior

No error.

The [eslint rule with the same name](http://eslint.org/docs/rules/no-invalid-this) allows for `this` references when assigning methods:

``` js
obj.foo = function foo() {
    // OK, this is in a method (this function assigns to a property).
    this.a = 0;
};
```

I would expect this to be valid in TypeScript as well.

Specificying an explicit `this` parameter does not clear the error.  I would consider this an acceptable alternative, if it worked.

**Comment by englercj** [comment]
Bumped up against this today while doing the following:

```js
MyClass.prototype.method = function () {
    return this.something; // `this` gives a tslint error for no-invalid-this
};
```

**Comment by adidahiya** [comment]
we're going to deprecate this rule; do you think you still need it given the compiler option? https://github.com/palantir/tslint/issues/1861

**Comment by errorx666** [comment]
No.

**Comment by englercj** [comment]
I feel like they are different. I'm OK with implicit `this` when on a class method, but want to disallow `this` (or more specifically, implicit `this`) when on a non-class method.

Right now there isn't a way to enforce that with just the compiler option, it is either all explicit or all implicit. Or do I misunderstand the usage?
